src/proxy-animation.js

```diff
@@ -30,151 +30,470 @@ function createReadyPromise(details) {
+  };
+
+  if (isFinished()) {
+    if (details.playState != 'finished') {
+      details.playState = 'finished';
+      // TODO: Ensure that finished promise and events fire.
+      // Possibly best to call finish after rAF though that would force a snap
+      // to the boundary, which is not quite correct.
+    }
+  } else if (details.playState == 'finished') {
+    details.playState = 'running';
+    addAnimation(details.timeline, details.proxy);
+  }
+}
+
+function hasActiveTimeline(details) {
+  return !details.timeline || details.timeline.phase != 'inactive';
+}
+
 // Create an alternate Animation class which proxies API requests.
 // TODO: Create a full-fledged proxy so missing methods are automatically
 // fetched from Animation.
 let proxyAnimations = new WeakMap();
+
 export class ProxyAnimation {
   constructor(effect, timeline) {
-    if (effect instanceof nativeAnimation) {
-      proxyAnimations.set(this, {
-        animation: effect,
-        timeline,
-        playState: "idle",
-        readyPromise: null,
-        sequence: 0, /* Used to track ready promises. */
-        aborted: new Set(), /* Aborted sequences. */
-      });
-      return;
-    }
-    const isScrollAnimation = timeline && timeline instanceof ScrollTimeline;
+    const animation =
+        (effect instanceof nativeAnimation) ?
+           effect : new nativeAnimation(effect, animationTimeline);
+    const isScrollAnimation = timeline instanceof ScrollTimeline;
     const animationTimeline = isScrollAnimation ? undefined : timeline;
     proxyAnimations.set(this, {
-      animation: new nativeAnimation(effect, animationTimeline),
+      animation: animation,
       timeline: isScrollAnimation ? timeline : undefined,
       playState: isScrollAnimation ? "idle" : null,
       readyPromise: null,
+      // Start and hold times are directly tracked in the proxy despite being
+      // accessible via the animation so that direct manipulation of these
+      // properties does not affect the play state of the underlying animation.
+      // Note that any changes to these values require an update of current
+      // time for the underlying animation to ensure that its hold time is set
+      // to the correct position.
+      startTime: null,
+      holdTime: null,
+      // When changing the timeline on a paused animation, we defer updating the
+      // start time until the animation resumes playing.
+      resetCurrentTimeOnResume: false,
+      // Calls to reverse and updatePlaybackRate set a pending rate that does
+      // not immediately take effect. The value of this property is
+      // inaccessible via the web animations API and therefore explicitly
+      // tracked.
+      pendingPlaybackRate: null,
+      proxy: this,
       sequence: 0, /* Used to track ready promises. */
       aborted: new Set(), /* Aborted sequences. */
     });
   }
 
+  // -----------------------------------------
+  // Helper method
+  // -----------------------------------------
```

Comment:
We shouldn't be exposing new methods on the animation object. Can we move this to an internal function like updateFinishedState? You can export this function to allow it to be called from other modules.

Reply:
Can A import something from B and B import something from A?

Reply:
Can use another design pattern to avoid adding a method to the animation object.